Fix logic for cors cache match

dlrobertson opened the pull request:
The current logic for a cors cache match does not consider "credentials is false and request's credentials mode is not "include" or credentials is true."

I could have missed something, but `CacheRequestDetails::credentials` is set to true if credentials mode is "include", and false otherwise. So `(!cors_cache.credentials && !cors_req.credentials) || cors_cache.credentials` would be directly following the spec, but unless I'm mistaken `cors_cache.credentials || !cors_req.credentials` is logically the same.

Fixes: #10525

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/10867)

<!-- Reviewable:end -->

dlrobertson committed:
Fix logic for cors cache match

The current logic for a cors cache match does not consider "credentials
is false and request's credentials mode is not "include" or credentials
is true."

highfive commented:
Heads up! This PR modifies the following files:
 * @kichjang: components/net/fetch/cors_cache.rs

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify net code, but no tests are modified. Please consider adding a test!

KiChjang commented on components/net/fetch/cors_cache.rs:
```diff
@@ -112,7 +112,7 @@ pub struct BasicCORSCache(Vec<CORSCacheEntry>);
 fn match_headers(cors_cache: &CORSCacheEntry, cors_req: &CacheRequestDetails) -> bool {
     cors_cache.origin == cors_req.origin &&
         cors_cache.url == cors_req.destination &&
```
Hmm... spec says "url is request's current url", so I'd expect to see something like `cors_cache.url == cors_req.current_url()` instead.

danlrobertson replied:
Great point. Didn't think to question that since the Issue was focused on the later logic. After looking at `CacheRequestDetails`, `destination` is the only `Url` member. Is `destination` used as the "current URL" in here?

KiChjang replied:
Wait, sorry. I didn't notice that `cors_req` is an instance of `CacheRequestDetails`, so the code as it is is fine.

danlrobertson replied:
No worries. Thanks for taking a peek at the code!

KiChjang commented:
@bors-servo r+

bors-servo commented:
:pushpin: Commit c01f8af has been approved by `KiChjang`

<!-- @bors-servo r=KiChjang c01f8af52f984e857a6645b7c33b3fc1c4f96724 -->

jdm commented:
@bors-servo: r-
Before we merge these changes, I'd like to either have a test for it or be explicit that we can't write one.

danlrobertson commented:
No problem. Where are the cors tests?

frewsxcv commented:
https://github.com/servo/servo/tree/master/tests/wpt/web-platform-tests/cors

jdm commented:
For the fetch code, it's `tests/unit/net/fetch.rs`. That code isn't hooked up to real WPT tests yet.

danlrobertson commented:
Review status: 0 of 4 files reviewed at latest revision, 2 unresolved discussions.

---

*[components/net/fetch/cors_cache.rs, line 153 \[r2\]](https://reviewable.io:443/reviews/servo/servo/10867#-KGNunTvNWNPoJZDixDp:-KGNunTw56wSRs84aBZL:837966876) ([raw file](https://github.com/servo/servo/blob/bc32e8e5ba413d4a1461d3b6ce6690666703fdf8/components/net/fetch/cors_cache.rs#L153)):*
`>` will find the entries that need to be dropped. `<` to find the ones to keep.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/10867)*
<!-- Sent from Reviewable.io -->

KiChjang commented on components/net/fetch/methods.rs:
```diff
@@ -40,6 +40,10 @@ pub fn fetch_async(request: Request, listener: Box<AsyncFetchListener + Send>) {
 
 /// [Fetch](https://fetch.spec.whatwg.org#concept-fetch)
 pub fn fetch(request: Rc<Request>) -> Response {
+    fetch_with_cors_cache(request, &mut BasicCORSCache::new())
```
This is not behaviour documented in the fetch spec. @jdm does this look ok?

danlrobertson replied:
Related to http://logs.glob.uno/?c=mozilla%23servo#c416083

jdm replied:
👍

KiChjang commented on components/net/fetch/methods.rs:
```diff
@@ -40,6 +40,10 @@ pub fn fetch_async(request: Request, listener: Box<AsyncFetchListener + Send>) {
+    fetch_with_cors_cache(request, &mut BasicCORSCache::new())
+}
+
+pub fn fetch_with_cors_cache<C: CORSCache>(request: Rc<Request>, cache: &mut C) -> Response {
```
I'm not sure why `cache: &mut CORSCache` wouldn't work here.

danlrobertson replied:
`CORSCache` is a trait, so you'll get the following error.
```
error: the trait bound `<trait>: std::marker::Sized` is not satisfied
```
I couldn't think of another way around it

dlrobertson committed:
Cleanup CORSCache

Remove the CORSCache trait, CORSCacheSender, CORSCacheThreadMsg, and
CORSCacheThread. Rename BasicCORSCache to CORSCache and keep its old
implementation of CORSCache.

KiChjang commented on components/net/fetch/cors_cache.rs:
```diff
@@ -74,51 +73,20 @@ pub struct CacheRequestDetails {
     pub credentials: bool
```
I'm wondering whether we even need this `CacheRequestDetails` struct at all... it feels like we can just do away this altogether and use `Request` instead. I think I'll file a follow-up issue to remove this.

KiChjang replied:
Filed #10904.

KiChjang commented:
@bors-servo r+

Thanks!

bors-servo commented:
:pushpin: Commit 483f07c has been approved by `KiChjang`

<!-- @bors-servo r=KiChjang 483f07c8f012ded73347dd0307eb4d09dba96a94 -->

bors-servo commented:
:pushpin: Commit 483f07c has been approved by `KiChjang`

<!-- @bors-servo r=KiChjang 483f07c8f012ded73347dd0307eb4d09dba96a94 -->

bors-servo commented:
:hourglass: Testing commit 483f07c with merge 3d38a60...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/592), [arm32](http://build.servo.org/builders/arm32/builds/628), [arm64](http://build.servo.org/builders/arm64/builds/613), [linux-dev](http://build.servo.org/builders/linux-dev/builds/624), [linux-rel](http://build.servo.org/builders/linux-rel/builds/672), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/640), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/621), [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/650)
